services/spar/src/Spar/Scim/User.hs

```diff
@@ -741,58 +846,98 @@ synthesizeStoredUser' uid veid dname handle richInfo accStatus createdAt lastUpd
+synthesizeScimUser tid info =
   let Handle userName = info ^. ST.vsuHandle
    in (Scim.empty ST.userSchemas userName (ST.ScimUserExtra (info ^. ST.vsuRichInfo)))
-        { Scim.externalId = Brig.renderValidExternalId $ info ^. ST.vsuExternalId,
+        { Scim.externalId = externalId tid (info ^. ST.vsuAuthId),
           Scim.displayName = Just $ fromName (info ^. ST.vsuName),
-          Scim.active = Just . Scim.ScimBool $ info ^. ST.vsuActive
+          Scim.active = Just . Scim.ScimBool $ info ^. ST.vsuActive,
+          Scim.emails =
+            maybeToList $
+              authIdEmailWithSource (info ^. ST.vsuAuthId) >>= \ews ->
+                case ewsEmailSource ews of
+                  EmailFromExternalIdField -> Nothing
+                  EmailFromEmailField -> emailToScimEmail (ewsEmail ews)
         }
+  where
+    externalId :: TeamId -> AuthId -> Maybe Text
+    externalId tid' authId = authIdExternalId tid' authId >>= \(ExternalId _ extId) -> pure extId
+
+    emailToScimEmail :: Email -> Maybe ScimEmail.Email
+    emailToScimEmail =
+      fmap scimEmail . toEmailAddress
+      where
+        scimEmail :: Text.Email.Validate.EmailAddress -> ScimEmail.Email
+        scimEmail addr = ScimEmail.Email Nothing (ScimEmail.EmailAddress2 addr) Nothing
 
 scimFindUserByHandle :: Maybe IdP -> TeamId -> Text -> MaybeT (Scim.ScimHandler Spar) (Scim.StoredUser ST.SparTag)
-scimFindUserByHandle mIdpConfig stiTeam hndl = do
+scimFindUserByHandle _mIdpConfig stiTeam hndl = do
   handle <- MaybeT . pure . parseHandle . Text.toLower $ hndl
   brigUser <- MaybeT . lift . Brig.getBrigUserByHandle $ handle
+
   guard $ userTeam (accountUser brigUser) == Just stiTeam
-  case Brig.veidFromBrigUser (accountUser brigUser) ((^. SAML.idpMetadata . SAML.edIssuer) <$> mIdpConfig) of
-    Right veid -> lift $ synthesizeStoredUser brigUser veid
-    Left _ -> Applicative.empty
 
--- | Construct a 'ValidExternalid'.  If it an 'Email', find the non-SAML SCIM user in spar; if
--- that fails, find the user by email in brig.  If it is a 'UserRef', find the SAML user.
+  -- TODO: remove authIdFromUser but keep same behaviour
+  case authIdFromUser (accountUser brigUser) of
+    Just _authId -> lift $ do
+      synthesizeStoredUser stiTeam brigUser
+    Nothing -> Applicative.empty
+
+-- | Find user by externalId, which is tried as a name id of a SAML User, an externalId spar,
+-- or an email in brig.
 -- Return the result as a SCIM user.
 --
 -- Note the user won't get an entry in `spar.user`.  That will only happen on their first
 -- successful authentication with their SAML credentials.
-scimFindUserByEmail :: Maybe IdP -> TeamId -> Text -> MaybeT (Scim.ScimHandler Spar) (Scim.StoredUser ST.SparTag)
-scimFindUserByEmail mIdpConfig stiTeam email = do
+scimFindUserByExternalId :: Maybe IdP -> TeamId -> Text -> MaybeT (Scim.ScimHandler Spar) (Scim.StoredUser ST.SparTag)
+scimFindUserByExternalId mIdpConfig stiTeam extId = do
```

Comment:
(This is where I'm giving up for today, can't focus any more.  I have read all of `git diff 0d4e837~..e038c923` except this function and changes under `spar/test-integration`.)